[C10D] Use future for flight recorder dump

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114817
* #114615
* __->__ #115176

Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Allows multiple calls to dump, which will be serialized.

Renames `tryWriteDebugInfo` to `launchAsyncDebugDump` in spirit of the
change to support more than one launch and to always launch rather than
only launching on the first call.

Adds a test for dumping on timeout.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @kiukchung @lucasllc

wconstab committed:
[C10D] Use future for flight recorder dump

Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Simplifies the test case for dump on timeout, which can now wait for
rank0 abort and assume the output file will be ready.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Use future for flight recorder dump"


Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Simplifies the test case for dump on timeout, which can now wait for
rank0 abort and assume the output file will be ready.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Use future for flight recorder dump"


Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Simplifies the test case for dump on timeout, which can now wait for
rank0 abort and assume the output file will be ready.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Use future for flight recorder dump"


Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Adds a test for dumping on timeout.

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -939,15 +939,33 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
 
-c10::optional<std::thread> ProcessGroupNCCL::tryWriteDebugInfo() {
+// TODO this launcher should be generalized, maybe stuck as a general 'c10'
+// helper
```
@zdevito i was thinking of leaving this for a follow up PR.  also do you have a suggestion where to put the util and what its interface should be?

i didn't do it here mainly to avoid figuring out the varargs and captures needed to pass args to the inner function from outside the launcher utility, but those should be solve-able with enough chat-gpting.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1066,7 +1084,9 @@ void ProcessGroupNCCL::dumpDebuggingInfo() {
       registerDebugInfoWriter(std::move(debugInfoWriterPtr));
     }
     debugInfoWriter_->write(ncclTrace);
+    return true;
```
annoyingly it looks like you need a real value for promise.setValue, so i went with a bool and it has some limited utility in showing which side of this if we landed on.

fduwjj replied:
I mean do we care about the return value or not?

wconstab replied:
i didn't bother checking it. i supposed its cleaner to actually check that it's true.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1280,15 +1300,9 @@ void ProcessGroupNCCL::workCleanupLoop() {
-                // becomes "best effort".
-                dumpingDebugInfo->detach();
-              }
+              auto& dumpingDebugInfo = tryWriteDebugInfo();
```
a little annoying, i couldn't stash the future above since its not copyable and i just took advantage of how tryWrite maintains a static.

If we don't want to keep it that way i can stash a pointer here to avoid multiple calls.

zdevito replied:
I think the lifetime of the return for tryWriteDebugInfo is a bit odd. Can we have it return a future, and then hold an optional future here if we need to have it not exist in certain cases?

wconstab replied:
yep, will do that

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -939,15 +939,33 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
+  return resultFuture;
+}
+
+std::future<bool>& ProcessGroupNCCL::tryWriteDebugInfo() {
```
this all dates back to the first impl, where we had a flag that prevented multiple attempts at dumping.

we still need some prevention, bc if you called this try function concurrently, you'd end up with races opening/writing the same file.

so keeping a static future was one way of ensuring only one thread, but allowing multiple callers to check the status of it and/or avoid having multiple callers result in multiple attempts.

zdevito replied:
I guess we can keep the mutex to ensure dumping is serialized (hold the mutex in the writer thread). I feel like multiple attempts is more the responsibility of the caller of the writing API.

wconstab commented on test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp:
```diff
@@ -188,10 +188,8 @@ class ProcessGroupNCCLNoHeartbeatCaught
-    if (thread) {
-      thread->join();
-    }
+    auto& fut = tryWriteDebugInfo();
```
well before, it would only write if it was the first. hence the 'try'.  but now i'm making it what you want.

wconstab committed:
Update on "[C10D] Use future for flight recorder dump"


Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Adds a test for dumping on timeout.

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Use future for flight recorder dump"


Replaces the "always sleep 30 sec before abort" with "wait up to 30 sec
for the future to complete then abort".  The difference in this case is
the abort happens as soon as the dump finishes up to a maximum, instead
of always waiting the maximum.

Allows multiple calls to dump, which will be serialized.

Renames `tryWriteDebugInfo` to `launchAsyncDebugDump` in spirit of the
change to support more than one launch and to always launch rather than
only launching on the first call.

Adds a test for dumping on timeout.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -939,15 +939,23 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
 
-c10::optional<std::thread> ProcessGroupNCCL::tryWriteDebugInfo() {
-  std::lock_guard<std::mutex> lock(writeDebugInfoMutex_);
-  if (writeDebugInfo_) {
```
Well this flag is still needed since we don't want the write is triggered multiple times?

fduwjj replied:
Well, without the gating of the flag, then inside monitor thread, after sleeping, it will still try to dump debug info. Or you are saying we should use two different file paths in two threads?

fduwjj replied:
I mean with the mutex's protection, the write will probably happen one by one, but since the file name is the same, I am afraid that we will overwrite the existing file.

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3699,6 +3702,78 @@ def gather_trace():
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(torch.cuda.device_count() < 2, "NCCL test requires 2+ GPUs")
+    def test_timeout_dumps(self):
+        if self.rank == self.MAIN_PROCESS_RANK:
```
The default value of `MAIN_PROCESS_RANK` is -1, IIUC, this if is actually an no-op?

wconstab replied:
no, main-process also runs this function, and its self.rank == -1.

so this lets us do some logic on the main process after the child processes exit

fduwjj replied:
OK, things I learned today... Upon reading the code, it is indeed working like what you mentioned..

wconstab commented:
@pytorchbot merge

huydhn commented:
@ptyorchbot revert -m 'Sorry for reverting your change, but the test_timeout_dumps is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/0e07e3dbe434ce31a5aea634628c7d39747f265f' -c weird

It's weird that PR jobs didn't fail

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but the test_timeout_dumps is failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/0e07e3dbe434ce31a5aea634628c7d39747f265f' -c weird

It's weird that PR jobs didn't fail

fduwjj commented:
@huydhn QQ, if this PR does not fail `test_timeout_dumps`, what made you suspect that this PR is the culprit?

huydhn commented:
Oh, it starts to fail in trunk after the PR is landed in trunk.  Here is the commit https://hud.pytorch.org/pytorch/pytorch/commit/0e07e3dbe434ce31a5aea634628c7d39747f265f.  The failure looks relevant to the change `FAILED [8.8153s] distributed/test_c10d_nccl.py::NCCLTraceTestDumpOnTimeout::test_timeout_dumps - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpfq8q3cdq/trace_0'`.  This could be a land race.

huydhn commented:
It is a bit easier to spot once you look at the distributed failures on HUD https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=distributed

fduwjj commented:
@huydhn that is interesting, failed test is the one added in this PR.. Is it because it is a new test, so CI does not capture it?

wconstab commented:
the failure on trunk looks like it is stable (it failed on 3 repeated tries) and for a consistent reason: we see the "preparing to dump" message but never see the "finished dump" message, so the file may indeed not be written by the time the test looks for it.

"""
2023-12-07T01:22:36.1092179Z distributed/test_c10d_nccl.py::NCCLTraceTestDumpOnTimeout::test_timeout_dumps [rank0]:[E ProcessGroupNCCL.cpp:530] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=2, OpType=ALLREDUCE, NumelIn=12, NumelOut=12, Timeout(ms)=1000) ran for 1691 milliseconds before timing out.
2023-12-07T01:22:36.1092654Z [rank0]:[E ProcessGroupNCCL.cpp:1075] ProcessGroupNCCL preparing to dump debug info.
2023-12-07T01:22:36.1093867Z [rank0]:[E ProcessGroupNCCL.cpp:544] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
2023-12-07T01:22:36.1094454Z [rank0]:[E ProcessGroupNCCL.cpp:550] To avoid data inconsistency, we are taking the entire process down.
2023-12-07T01:22:36.1096221Z [rank0]:[E ProcessGroupNCCL.cpp:1204] [Rank 0] NCCL watchdog thread terminated with exception: [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=2, OpType=ALLREDUCE, NumelIn=12, NumelOut=12, Timeout(ms)=1000) ran for 1691 milliseconds before timing out.
2023
"""

This is within our design parameters, if the dumping process is taking a longer than expected time, we intentionally exit before waiting for it.  So it's possible we need to wait longer, but i'm not sure why it's necessary in particular on trunk.

In the past the dump function took longer due to slow addr2line.  However we should have disabled this by not capturing CPP logs.  (TORCH_NCCL_TRACE_CPP_STACK=False).  So i am not expecting a slow addr2line on the machine is relevant.

Will rebase to see if there is indeed a land race.

We

wconstab commented:
Ok i put up a reland PR with just the changes to make the test code line up with @zdevito's dump-pipe PR.  I can't repro the 'took 50 sec' error, its taking 24 sec for me on several repeats now.

Also @fduwjj I temporarily put extra logs into the watchdog thread and the heartbeat monitor to observe what happens on shutdown.  The watchdog thread performs the dump, waits up to 30sec, then finally raises the timeout exception.  This exception kills the process, which prevents the heartbeatMonitor thread from ever trying to issue a dump.  So IIUC this should be the normal case.  Only in the case where destructors hang during exception handling, should the process fail to terminate in this way and then we'd expect a second dump and an abort to come from the heartbeat monitor.

fduwjj commented:
> I temporarily put extra logs into the watchdog thread and the heartbeat monitor to observe what happens on shutdown. The watchdog thread performs the dump, waits up to 30sec, then finally raises the timeout exception. This exception kills the process, which prevents the heartbeatMonitor thread from ever trying to issue a dump.

So you are saying that when watchdog throws timeout, it will just directly kill the process so that heartbeatMonitor thread will be killed as well so the dump will not be triggered?
